background-charm-service/src/worker-controller.ts

```diff
@@ -44,118 +64,128 @@ export class WorkerController {
-      err.preventDefault();
-    };
+    this.worker.addEventListener("message", this.onWorkerMessage);
+    this.worker.addEventListener("error", this.onWorkerError);
   }
 
   async initialize() {
-    if (this.initDeferred) {
-      return this.initDeferred.promise;
+    if (this.state !== WorkerState.Uninitialized) {
+      throw new Error("Worker is not uninitialized.");
     }
-    this.initDeferred = defer();
+    this.state = WorkerState.Initializing;
     try {
-      await this.exec("setup", {
+      await this.exec(WorkerIPCMessageType.Initialize, {
         did: this.did,
         toolshedUrl: this.toolshedUrl,
         rawIdentity: this.identity.serialize(),
       });
-      this.initDeferred.resolve();
-      this.ready = true;
+      this.state = WorkerState.Ready;
     } catch (e) {
-      this.initDeferred.reject(
-        new Error(`Failed to initialize Worker: ${e ? e.toString() : e}`),
-      );
+      this.state = WorkerState.Error;
+      throw e;
     }
-    return this.initDeferred;
   }
 
-  runCharm(
+  async runCharm(
     bg: Cell<BGCharmEntry>,
-  ): Promise<{ success: boolean; data?: any; charmId: string }> {
-    return this.exec("runCharm", { charmId: bg.get().charmId });
+  ): Promise<void> {
+    if (this.state !== WorkerState.Ready) {
+      throw new Error("Worker not ready.");
+    }
+    return await this.exec(WorkerIPCMessageType.Run, {
+      charmId: bg.get().charmId,
+    });
   }
 
   async shutdown() {
+    if (
+      this.state === WorkerState.Terminating ||
+      this.state === WorkerState.Terminated
+    ) {
+      throw new Error(`Worker is already ${this.state}.`);
+    }
+    this.state = WorkerState.Terminating;
+
+    for (const [_, deferred] of this.pending.entries()) {
+      deferred.reject(new Error("Worker shutting down."));
+    }
+    this.pending.clear();
+
     try {
-      await this.exec("shutdown");
+      await this.exec(WorkerIPCMessageType.Cleanup);
     } catch (err) {
-      log(
-        "Failed to shutdown worker gracefully, terminating with unknown status.",
+      console.warn(
+        `Failed to shutdown worker gracefully: ${err}`,
       );
-      this.worker?.terminate();
     }
+    this.worker.terminate();
+    this.state = WorkerState.Terminated;
+  }
+
+  isReady(): boolean {
+    return this.state === WorkerState.Ready;
   }
 
   // send a message and return a promise that resolves with the response
-  private exec(type: string, data?: any): Promise<any> {
-    // Only allow "setup" calls when not ready
-    if (type !== "setup" && !this.ready) {
-      return Promise.reject(new Error("Worker not initialized"));
-    }
+  private exec(type: WorkerIPCMessageType, data?: any): Promise<void> {
     const msgId = this.msgId++;
+    const message: WorkerIPCRequest = {
+      msgId,
+      type,
+      data,
+    };
 
-    const deferred = defer<any, Error>();
+    const deferred = defer<void, Error>();
 
     const timeout = setTimeout(() => {
```

Comment:
moving the only timeout for this to here is great - it will allow us to have different timeouts for different parts of the sequence if we want!  (not that we need it now)